SNP/SNPModule.cpp

```diff
@@ -222,70 +173,74 @@ static BOOL __stdcall spi_lock_game_list(int, int, game** out_game_list) {
 
         if (g_snp_context.game_app_info.version_id != ad.game_info.version_id) {
             joinable = false;
-            prefixes += std::format("{}", char(ColorByte::Gray));
-            prefixes += "[!Ver] ";
-            prefixes += char(ColorByte::Gray);
+            ss << ColorByte::Gray << "[!Ver]";
         }
-
         if (snp_config.mode != ad.crownlink_mode) {
             joinable = false;
-            prefixes += std::format("{}", char(ColorByte::Gray));
-            prefixes += "[!Mode] ";
+            ss << ColorByte::Gray << "[!Mode]";
         }
-
         if (ad == g_snp_context.status_ad) {
             joinable = false;
         }
-
         if (ad.game_info.game_state == 12) {
-            // game in progress - these won't be shown anyway but we don't want to initiate p2p
+            // Game in progress - these won't be shown anyway but we don't want to initiate p2p
             joinable = false;
         }
 
-        // agent_state calls ensure_agent so one will be created if needed
+        // agent_state calls ensure_agent so one will be created if needed,
         // this kicks off the p2p connection process
         if (joinable) {
             switch (g_crown_link->juice_manager().lobby_agent_state(ad)) {
                 case JUICE_STATE_CONNECTING: {
-                    prefixes += "...";
-                    prefixes += std::format("{}", char(ColorByte::Gray));
+                    ss << ColorByte::Gray << "...";
                 } break;
                 case JUICE_STATE_FAILED: {
-                    prefixes += " !! ";
-                    prefixes += std::format("{}", char(ColorByte::Gray));
+                    ss << ColorByte::Gray << " !! ";
                     g_crown_link->juice_manager().send_connection_request(ad.game_info.host);
                 } break;
                 case JUICE_STATE_DISCONNECTED: {
-                    prefixes += "-";
-                    prefixes += std::format("{}", char(ColorByte::Gray));
+                    ss << ColorByte::Gray << "-";
                     g_crown_link->juice_manager().send_connection_request(ad.game_info.host);
                 } break;
                 case JUICE_STATE_CONNECTED:
                 case JUICE_STATE_COMPLETED: {
                     switch (g_crown_link->juice_manager().final_connection_type(ad.game_info.host)) {
                         case JuiceConnectionType::Relay: {
-                            prefixes += "[R]";
+                            ss << ColorByte::Gray << "[R]";
                         } break;
                         case JuiceConnectionType::Radmin: {
-                            prefixes += std::format("[Radmin {}]", char(131));
+                            ss << ColorByte::Gray << std::format("[Radmin {}]", static_cast<char>(131));  // char :(
+                        } break;
+                        default: {
+                            ss << ColorByte::Gray << static_cast<char>(187);  // char >>
```

Comment:
The states in this switch/case are for lobbies that we can join - they should not be gray. There should be no ColorByte added.

Reply:
What ColorByte should their have then? To handle colors robustly it's best to always set some color on each fragment, to prevent leaks from previous writes

Reply:
No, it needs to be empty.

When there's no color byte set starcraft treats them as menu items where they're dark green until you click on them then they are highlighted in light green. This is important for players who may use their keyboard to navigate the list as the light green highlighted lobby is the one that's currently selected and otherwise they have no way of knowing where they are in the list.

It would be cool if we could extend that behavior in CM and also highlight the gray ones (e.g. gray -> white) when they're selected but that's irrelevant to the current behavior. Best we can do now is leave it blank so the selection behavior remains.

By the way `0x1` isn't default, it's "revert to previous color" so by setting that there the lobbies "revert" to the blue color used in the games screen:
![image](https://github.com/user-attachments/assets/d70f6cb8-9989-4edb-be8a-5b0d2842c097)

Reply:
Also: here's how the `joinable == false` lobbies look with your changes:
![image](https://github.com/user-attachments/assets/c50ff767-6444-411f-8541-26f18c0f1b7c)

Because in this case you're appending the `ColorByte::Default` which is really reverting to previous it's turning the map name back to green.

Reply:
Here you go - here's what you should do:

```
static BOOL __stdcall spi_lock_game_list(int, int, GameInfo** out_game_list) {
    // called by storm once per second when on the games list screen
    g_advertisement_mutex.lock();

    // storm will unlock when it's done with the data by calling spi_unlock_game_list()
    const auto& snp_config = SnpConfig::instance();

    s32 game_index = 1;
    AdFile* last_ad = nullptr;

    update_status_ad();
    last_ad = &g_snp_context.status_ad;

    std::erase_if(g_snp_context.lobbies, [](AdFile lobby) {
        return get_tick_count() - lobby.game_info.host_last_time > 3;
    });

    for (AdFile& ad : g_snp_context.lobbies) {
        bool joinable = true;
        std::stringstream ss;

        if (g_snp_context.game_app_info.version_id != ad.game_info.version_id) {
            joinable = false;
            ss << ColorByte::Red << "[!Ver]";
        }
        if (snp_config.mode != ad.crownlink_mode) {
            joinable = false;
            ss << ColorByte::Red << "[!Mode]";
        }
        if (ad == g_snp_context.status_ad) {
            // Status ad is set to 0x2 (blue) in part to make sure it's at the top
            // If we set it to red it'll be below the lobbies not above
            joinable = false;
            ss << ColorByte::Blue;
        }
        if (ad.game_info.game_state == 12) {
            // Game in progress - these won't be shown anyway but we don't want to initiate p2p
            joinable = false;
            ss << ColorByte::Red;
        }

        // agent_state calls ensure_agent so one will be created if needed,
        // this kicks off the p2p connection process
        if (joinable) {
            switch (g_crown_link->juice_manager().lobby_agent_state(ad)) {
                case JUICE_STATE_CONNECTING: {
                    ss << ColorByte::Gray << "...";
                } break;
                case JUICE_STATE_FAILED: {
                    ss << ColorByte::Gray << "!!";
                    g_crown_link->juice_manager().send_connection_request(ad.game_info.host);
                } break;
                case JUICE_STATE_DISCONNECTED: {
                    ss << ColorByte::Gray << "-";
                    g_crown_link->juice_manager().send_connection_request(ad.game_info.host);
                } break;
                case JUICE_STATE_CONNECTED:
                case JUICE_STATE_COMPLETED: {
                    switch (g_crown_link->juice_manager().final_connection_type(ad.game_info.host)) {
                        case JuiceConnectionType::Relay: {
                            ss << ColorByte::Green << "[R]";
                        } break;
                        case JuiceConnectionType::Radmin: {
                            ss << ColorByte::Green << std::format("[Radmin {}]", static_cast<char>(131));  // char :(
                        } break;
                        default: {
                            ss << ColorByte::Green << static_cast<char>(187);  // char >>
                        } break;
                    }
                    ss << ColorByte::Revert; // Go back to "menu" behavior after first character
                    // This makes the lobbies dark green then turn light green when selected
                }
            }
        }

        if (ad.original_name.empty()) {
            ad.original_name = ad.game_info.game_name;
        }
        ss << " " << ad.original_name;

        if (snp_config.add_map_to_lobby_name && ad != g_snp_context.status_ad) {
            ss << (joinable ? ColorByte::Blue : ColorByte::Gray) << " " << extract_map_name(ad.game_info);
        }

        const auto prefixes = ss.str();
        spdlog::debug("Lobby name updated to {}, length: {}", prefixes, prefixes.size());
        strncpy(ad.game_info.game_name, prefixes.c_str(), sizeof(ad.game_info.game_name));

        if (last_ad) {
            last_ad->game_info.pNext = &ad.game_info;
        }
        last_ad = &ad;
    }
```